q2_longitudinal/tests/test_longitudinal.py

```diff
@@ -288,6 +310,149 @@ def test_nmit_missing_table_ids(self):
+                        name='Difference')
+        exp.index.name = '#SampleID'
+        obs = first_differences(
+            metadata=qiime2.Metadata(md_static), state_column='Time',
+            individual_id_column='ind',
+            metric='Value', replicate_handling='drop')
+        pdt.assert_series_equal(obs.sort_index(), exp.sort_index())
+
+    def test_first_differences_drop_duplicates(self):
+        obs = first_differences(
+            metadata=qiime2.Metadata(md_dup), state_column='Time',
+            individual_id_column='ind',
+            metric='Value', replicate_handling='random')
+        # The first diff of individual 2 is subject to random rep handling
+        mystery_number = obs.iloc[1]
+        self.assertTrue(mystery_number in [0.04999999999999999, 0.06])
+        # but other values are constant, so we will just drop in the mystery
+        # value and the exp/obs series should match.
+        exp = pd.Series([0.08, mystery_number, 0.12, 0.14,
+                         0.14999999999999997],
+                        index=['3', '4', '9', '10', '11'], name='Difference')
+        exp.index.name = '#SampleID'
+        pdt.assert_series_equal(obs.sort_index(), exp.sort_index())
+
+    def test_first_differences_single_state(self):
+        single_state = qiime2.Metadata(md[md['Time'] == 1])
+        with self.assertRaisesRegex(ValueError, "state_column must contain"):
+            first_differences(
+                metadata=single_state, state_column='Time',
+                individual_id_column='ind',
+                metric='Value', replicate_handling='drop')
+
+    def test_first_differences_single_individual(self):
+        exp = pd.Series([0.08],
+                        index=['3'],
+                        name='Difference')
+        exp.index.name = '#SampleID'
+        single_ind = qiime2.Metadata(md[md['ind'] == 1])
+        obs = first_differences(
+                metadata=single_ind, state_column='Time',
+                individual_id_column='ind',
+                metric='Value', replicate_handling='drop')
+        pdt.assert_series_equal(obs.sort_index(), exp.sort_index())
+
+    def test_first_differences_single_sample(self):
+        single_sam = qiime2.Metadata(md[(md['ind'] == 1) & (md['Time'] == 1)])
+        with self.assertRaisesRegex(ValueError, "state_column must contain"):
+            first_differences(
+                metadata=single_sam, state_column='Time',
+                individual_id_column='ind',
+                metric='Value', replicate_handling='drop')
+
+    def test_first_differences_empty(self):
+        with self.assertRaisesRegex(ValueError, "Metadata is empty"):
+            first_differences(
+                metadata=(qiime2.Metadata(pd.DataFrame({'a': [], 'b': []}))),
+                state_column='Time', individual_id_column='ind',
+                metric='Value', replicate_handling='drop')
+
+    def test_first_differences_taxa(self):
+        first_differences(
```

Comment:
Can you assert the results are correct here?

Reply:
👍 I've added the expected results as a tsv